File: scripts/discover_status_pages.py
#!/usr/bin/env python3
"""Script to discover lift status pages for ski resorts."""

import argparse
import asyncio
import csv
import json
import sys
from pathlib import Path

import structlog

# Add src to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from ski_lift_status import load_resorts, load_lifts, get_lifts_for_resort
from ski_lift_status.scraping.discovery import DiscoveryAgent, DiscoveryResult

logger = structlog.get_logger()

DATA_DIR = Path(__file__).parent.parent / "data"
STATUS_PAGES_FILE = DATA_DIR / "status_pages.csv"


def load_existing_status_pages() -> dict[str, dict]:
    """Load existing status pages CSV."""
    existing = {}
    if STATUS_PAGES_FILE.exists():
        with open(STATUS_PAGES_FILE, newline="", encoding="utf-8") as f:
            reader = csv.DictReader(f)
            for row in reader:
                existing[row["resort_id"]] = row
    return existing


def save_status_pages(entries: list[dict]) -> None:
    """Save status pages to CSV."""
    fieldnames = ["resort_id", "resort_name", "website_url", "status_page_url"]

    with open(STATUS_PAGES_FILE, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        for entry in entries:
            writer.writerow(entry)


def get_top_resorts_by_lifts(resorts, lifts, top_n: int = 30) -> list:
    """Get top N resorts by number of lifts.

    Args:
        resorts: List of Resort objects.
        lifts: List of all Lift objects.
        top_n: Number of top resorts to return.

    Returns:
        List of (resort, lift_count) tuples sorted by lift count.
    """
    resort_lift_counts = {}

    for lift in lifts:
        if not lift.ski_area_ids:
            continue
        for ski_area_id in lift.ski_area_ids.split(";"):
            ski_area_id = ski_area_id.strip()
            if ski_area_id:
                resort_lift_counts[ski_area_id] = resort_lift_counts.get(ski_area_id, 0) + 1

    # Create resort lookup
    resort_by_id = {r.id: r for r in resorts}

    # Sort by lift count
    sorted_resorts = sorted(
        [(resort_by_id.get(rid), count) for rid, count in resort_lift_counts.items()
         if resort_by_id.get(rid)],
        key=lambda x: x[1],
        reverse=True,
    )

    return sorted_resorts[:top_n]


async def discover_status_page(
    resort_id: str,
    resort_name: str,
    resort_website: str | None,
    lifts: list,
) -> DiscoveryResult:
    """Discover status page for a single resort."""
    agent = DiscoveryAgent()

    lift_names = [lift.name for lift in lifts if lift.name]

    return await agent.discover(
        resort_id=resort_id,
        resort_name=resort_name,
        lift_names=lift_names,
        resort_website=resort_website,
    )


async def run_discovery(
    resort_ids: list[str] | None = None,
    top_n: int | None = None,
    override: bool = False,
    output_json: str | None = None,
) -> list[DiscoveryResult]:
    """Run discovery for resorts.

    Args:
        resort_ids: Specific resort IDs to discover. If None, uses top_n.
        top_n: Number of top resorts by lifts to discover.
        override: If True, re-discover even if already in CSV.
        output_json: Path to write JSON results.

    Returns:
        List of DiscoveryResult objects.
    """
    resorts = load_resorts()
    all_lifts = load_lifts()
    existing = load_existing_status_pages()

    resort_by_id = {r.id: r for r in resorts}

    # Determine which resorts to process
    resorts_to_process = []

    if resort_ids:
        # Specific resorts requested
        for rid in resort_ids:
            if rid in resort_by_id:
                resort = resort_by_id[rid]
                if not override and rid in existing:
                    logger.info("skipping_existing", resort_id=rid, resort_name=resort.name)
                    continue
                resorts_to_process.append(resort)
            else:
                logger.warning("resort_not_found", resort_id=rid)
    else:
        # Use top N resorts
        n = top_n or 30
        top_resorts = get_top_resorts_by_lifts(resorts, all_lifts, n)

        for resort, lift_count in top_resorts:
            if not override and resort.id in existing:
                logger.info("skipping_existing", resort_id=resort.id, resort_name=resort.name)
                continue
            resorts_to_process.append(resort)

    logger.info("resorts_to_process", count=len(resorts_to_process))

    # Run discovery
    results = []
    for resort in resorts_to_process:
        lifts = get_lifts_for_resort(resort.id)
        website = resort.websites.split(";")[0].strip() if resort.websites else None

        logger.info(
            "discovering",
            resort_id=resort.id,
            resort_name=resort.name,
            lift_count=len(lifts),
        )

        try:
            result = await discover_status_page(
                resort_id=resort.id,
                resort_name=resort.name,
                resort_website=website,
                lifts=lifts,
            )
            results.append(result)

            if result.success:
                logger.info(
                    "discovered",
                    resort_name=resort.name,
                    url=result.status_page_url,
                    confidence=result.confidence,
                )
            else:
                logger.warning(
                    "discovery_failed",
                    resort_name=resort.name,
                    errors=result.errors,
                )

        except Exception as e:
            logger.error("discovery_error", resort_name=resort.name, error=str(e))
            results.append(DiscoveryResult(
                resort_id=resort.id,
                resort_name=resort.name,
                success=False,
                errors=[str(e)],
            ))

        # Rate limiting
        await asyncio.sleep(1.0)

    # Update status_pages.csv with new entries
    all_entries = []

    # Keep existing entries
    for rid, row in existing.items():
        # Skip if we're overriding this resort
        if override and any(r.resort_id == rid for r in results):
            continue
        all_entries.append(row)

    # Add new successful discoveries
    for result in results:
        if result.success and result.status_page_url:
            all_entries.append({
                "resort_id": result.resort_id,
                "resort_name": result.resort_name,
                "website_url": result.website_url or "",
                "status_page_url": result.status_page_url,
            })

    # Sort by resort name
    all_entries.sort(key=lambda x: x["resort_name"])

    # Save updated CSV
    save_status_pages(all_entries)
    logger.info("saved_status_pages", count=len(all_entries))

    # Output JSON if requested
    if output_json:
        json_results = []
        for r in results:
            json_results.append({
                "resort_id": r.resort_id,
                "resort_name": r.resort_name,
                "success": r.success,
                "status_page_url": r.status_page_url,
                "website_url": r.website_url,
                "confidence": r.confidence,
                "reasoning": r.reasoning,
                "search_queries": r.search_queries,
                "candidate_urls": r.candidate_urls,
                "errors": r.errors,
            })

        with open(output_json, "w") as f:
            json.dump({
                "results": json_results,
                "total": len(results),
                "successful": sum(1 for r in results if r.success),
                "failed": sum(1 for r in results if not r.success),
            }, f, indent=2)

        logger.info("saved_json_results", path=output_json)

    return results


def main():
    parser = argparse.ArgumentParser(
        description="Discover lift status pages for ski resorts"
    )
    parser.add_argument(
        "--resort-id",
        type=str,
        help="Specific resort ID to discover",
    )
    parser.add_argument(
        "--resort-ids",
        type=str,
        help="Comma-separated list of resort IDs",
    )
    parser.add_argument(
        "--top",
        type=int,
        default=30,
        help="Number of top resorts by lift count (default: 30)",
    )
    parser.add_argument(
        "--all",
        action="store_true",
        help="Discover for all resorts",
    )
    parser.add_argument(
        "--override",
        action="store_true",
        help="Override existing entries in status_pages.csv",
    )
    parser.add_argument(
        "--output",
        type=str,
        help="Path to write JSON results",
    )

    args = parser.parse_args()

    # Configure logging
    structlog.configure(
        wrapper_class=structlog.make_filtering_bound_logger(logging.INFO),
    )

    # Determine resort IDs
    resort_ids = None
    if args.resort_id:
        resort_ids = [args.resort_id]
    elif args.resort_ids:
        resort_ids = [rid.strip() for rid in args.resort_ids.split(",")]
    elif args.all:
        # Load all resorts
        resorts = load_resorts()
        resort_ids = [r.id for r in resorts]

    # Run discovery
    results = asyncio.run(run_discovery(
        resort_ids=resort_ids,
        top_n=args.top if not resort_ids else None,
        override=args.override,
        output_json=args.output,
    ))

    # Print summary
    successful = sum(1 for r in results if r.success)
    failed = sum(1 for r in results if not r.success)

    print(f"\n{'='*60}")
    print(f"Discovery Complete")
    print(f"{'='*60}")
    print(f"Total processed: {len(results)}")
    print(f"Successful: {successful}")
    print(f"Failed: {failed}")

    if successful > 0:
        print(f"\nDiscovered status pages:")
        for r in results:
            if r.success:
                print(f"  - {r.resort_name}: {r.status_page_url}")
                print(f"    Confidence: {r.confidence:.1%}")


if __name__ == "__main__":
    import logging
    main()

File: scripts/discover-status-pages.ts
#!/usr/bin/env tsx
/**
 * Discover Status Pages
 * Finds lift/run status pages for ski resorts using AI
 */

import * as https from 'https';
import * as fs from 'fs';
import * as path from 'path';

const DATA_DIR = path.join(__dirname, '../data');
const SERPER_API_KEY = process.env.SERPER_API_KEY;
const OPENAI_API_KEY = process.env.OPENAI_API_KEY;

interface Resort {
  id: string;
  name: string;
  countries: string;
  websites: string;
}

interface SearchResult {
  title: string;
  link: string;
  snippet: string;
}

interface DiscoveryResult {
  resort_id: string;
  resort_name: string;
  success: boolean;
  status_page_url: string | null;
  confidence: number;
}

/**
 * Make HTTPS request
 */
async function httpsRequest(options: https.RequestOptions, data?: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let body = '';
      res.on('data', (chunk) => (body += chunk));
      res.on('end', () => {
        if (res.statusCode && res.statusCode >= 200 && res.statusCode < 300) {
          resolve(body);
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${body}`));
        }
      });
    });

    req.on('error', reject);
    if (data) req.write(data);
    req.end();
  });
}

/**
 * Search for resort status pages using Serper
 */
async function searchStatusPage(resortName: string, website: string): Promise<SearchResult[]> {
  if (!SERPER_API_KEY) {
    throw new Error('SERPER_API_KEY environment variable is required');
  }

  const query = `${resortName} lift status piste map live`;
  const searchData = JSON.stringify({ q: query, num: 5 });

  const options: https.RequestOptions = {
    hostname: 'google.serper.dev',
    path: '/search',
    method: 'POST',
    headers: {
      'X-API-KEY': SERPER_API_KEY,
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(searchData),
    },
  };

  const response = await httpsRequest(options, searchData);
  const data = JSON.parse(response);

  return (data.organic || []).map((result: any) => ({
    title: result.title,
    link: result.link,
    snippet: result.snippet,
  }));
}

/**
 * Analyze search results using OpenAI
 */
async function analyzeResults(
  resortName: string,
  results: SearchResult[]
): Promise<{ url: string | null; confidence: number }> {
  if (!OPENAI_API_KEY) {
    throw new Error('OPENAI_API_KEY environment variable is required');
  }

  const prompt = `You are analyzing search results to find the official lift/run status page for "${resortName}" ski resort.

Search results:
${results.map((r, i) => `${i + 1}. ${r.title}\n   URL: ${r.link}\n   ${r.snippet}`).join('\n\n')}

Which URL is most likely the official live lift/run status page? Respond with ONLY a JSON object:
{
  "url": "the best URL or null if none found",
  "confidence": 0.0-1.0,
  "reason": "brief explanation"
}`;

  const requestData = JSON.stringify({
    model: 'gpt-4o-mini',
    messages: [
      {
        role: 'system',
        content: 'You are a helpful assistant that identifies official ski resort status pages.',
      },
      { role: 'user', content: prompt },
    ],
    temperature: 0.3,
    response_format: { type: 'json_object' },
  });

  const options: https.RequestOptions = {
    hostname: 'api.openai.com',
    path: '/v1/chat/completions',
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${OPENAI_API_KEY}`,
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(requestData),
    },
  };

  const response = await httpsRequest(options, requestData);
  const data = JSON.parse(response);

  const content = data.choices[0].message.content;
  const analysis = JSON.parse(content);

  return {
    url: analysis.url,
    confidence: analysis.confidence,
  };
}

/**
 * Load resorts from CSV
 */
function loadResorts(limit?: number): Resort[] {
  const csvPath = path.join(DATA_DIR, 'resorts.csv');
  const content = fs.readFileSync(csvPath, 'utf-8');
  const lines = content.split('\n').filter((l) => l.trim());

  const resorts: Resort[] = [];
  for (let i = 1; i < lines.length; i++) {
    // Skip header
    const parts = lines[i]!.split(',');
    if (parts.length >= 4) {
      resorts.push({
        id: parts[0]!,
        name: parts[1]!,
        countries: parts[2]!,
        websites: parts[5] || '',
      });
    }

    if (limit && resorts.length >= limit) break;
  }

  return resorts;
}

/**
 * Save results to CSV
 */
function saveResults(results: DiscoveryResult[], outputPath: string) {
  const csv = [
    'resort_id,resort_name,status_page_url,confidence',
    ...results.map(
      (r) =>
        `${r.resort_id},"${r.resort_name}",${r.status_page_url || ''},${r.confidence.toFixed(2)}`
    ),
  ].join('\n');

  fs.writeFileSync(outputPath, csv, 'utf-8');
}

/**
 * Main discovery function
 */
async function discover(options: { top?: number; all?: boolean; resortId?: string; override?: boolean }) {
  console.log('🔍 Starting status page discovery...\n');

  let resorts: Resort[];

  if (options.resortId) {
    const allResorts = loadResorts();
    const resort = allResorts.find((r) => r.id === options.resortId);
    if (!resort) {
      throw new Error(`Resort not found: ${options.resortId}`);
    }
    resorts = [resort];
  } else if (options.all) {
    resorts = loadResorts();
  } else {
    resorts = loadResorts(options.top || 30);
  }

  console.log(`Processing ${resorts.length} resorts...\n`);

  const results: DiscoveryResult[] = [];

  for (let i = 0; i < resorts.length; i++) {
    const resort = resorts[i]!;
    console.log(`[${i + 1}/${resorts.length}] ${resort.name}`);

    try {
      // Search for status page
      const searchResults = await searchStatusPage(resort.name, resort.websites);

      if (searchResults.length === 0) {
        console.log('  ❌ No search results found\n');
        results.push({
          resort_id: resort.id,
          resort_name: resort.name,
          success: false,
          status_page_url: null,
          confidence: 0,
        });
        continue;
      }

      // Analyze with AI
      const analysis = await analyzeResults(resort.name, searchResults);

      if (analysis.url) {
        console.log(`  ✅ Found: ${analysis.url} (${(analysis.confidence * 100).toFixed(0)}%)\n`);
        results.push({
          resort_id: resort.id,
          resort_name: resort.name,
          success: true,
          status_page_url: analysis.url,
          confidence: analysis.confidence,
        });
      } else {
        console.log('  ❌ No suitable page found\n');
        results.push({
          resort_id: resort.id,
          resort_name: resort.name,
          success: false,
          status_page_url: null,
          confidence: 0,
        });
      }

      // Rate limiting
      await new Promise((resolve) => setTimeout(resolve, 1000));
    } catch (error) {
      console.log(`  ❌ Error: ${error}\n`);
      results.push({
        resort_id: resort.id,
        resort_name: resort.name,
        success: false,
        status_page_url: null,
        confidence: 0,
      });
    }
  }

  // Save results
  const outputPath = path.join(DATA_DIR, 'status_pages.csv');
  saveResults(results, outputPath);

  const jsonPath = 'discovery-results.json';
  const summary = {
    total: results.length,
    successful: results.filter((r) => r.success).length,
    failed: results.filter((r) => !r.success).length,
    results,
  };
  fs.writeFileSync(jsonPath, JSON.stringify(summary, null, 2));

  console.log('\n📊 Summary:');
  console.log(`  Total: ${summary.total}`);
  console.log(`  Successful: ${summary.successful}`);
  console.log(`  Failed: ${summary.failed}`);
  console.log(`\n✅ Results saved to ${outputPath}`);
  console.log(`📄 JSON output: ${jsonPath}`);
}

/**
 * CLI
 */
async function main() {
  const args = process.argv.slice(2);

  const options = {
    top: args.includes('--top') ? parseInt(args[args.indexOf('--top') + 1] || '30') : undefined,
    all: args.includes('--all'),
    resortId: args.includes('--resort-id') ? args[args.indexOf('--resort-id') + 1] : undefined,
    override: args.includes('--override'),
  };

  // Validate API keys
  if (!SERPER_API_KEY) {
    console.error('❌ SERPER_API_KEY environment variable is required');
    process.exit(1);
  }

  if (!OPENAI_API_KEY) {
    console.error('❌ OPENAI_API_KEY environment variable is required');
    process.exit(1);
  }

  await discover(options);
}

if (require.main === module) {
  main().catch((error) => {
    console.error('Error:', error);
    process.exit(1);
  });
}

export { discover };
